src/openfermion/utils/_commutator_diagonal_coulomb_operator.py

```diff
@@ -0,0 +1,303 @@
+                warnings.warn('Defaulted to standard commutator evaluation '
+                              'due to an out-of-spec operator.')
+                additional = FermionOperator.zero()
+                additional.terms[term_a + term_b] = coefficient
+                additional.terms[term_b + term_a] = -coefficient
+                additional = normal_ordered(additional)
+
+                result += additional
+
+    return result
+
+
+def _commutator_one_body_with_one_body(one_body_action_a, one_body_action_b,
+                                       coefficient, result):
+    """Compute the commutator of two one-body operators specified by actions.
+
+    Args:
+        one_body_action_a, one_body_action_b (tuple): single terms of one-body
+            FermionOperators (i^ j or i^ i).
+        coefficient (complex float): coefficient of the commutator.
+        result (FermionOperator): the resultant to which to add the commutator.
+    """
+    # In the case that both the creation and annihilation operators of the
+    # two actions pair, two new terms must be added.
+    if (one_body_action_a[0][0] == one_body_action_b[1][0] and
+            one_body_action_b[0][0] == one_body_action_a[1][0]):
+        new_one_body_action_a = ((one_body_action_a[0][0], 1),
+                                 (one_body_action_a[0][0], 0))
+        new_one_body_action_b = ((one_body_action_b[0][0], 1),
+                                 (one_body_action_b[0][0], 0))
+
+        result.terms[new_one_body_action_a] = (
+            result.terms.get(new_one_body_action_a, 0.0) + coefficient)
+        result.terms[new_one_body_action_b] = (
+            result.terms.get(new_one_body_action_b, 0.0) - coefficient)
+
+    # A single pairing causes the mixed action a[0]^ b[1] to be added
+    elif one_body_action_a[1][0] == one_body_action_b[0][0]:
+        action_ab = ((one_body_action_a[0][0], 1),
+                     (one_body_action_b[1][0], 0))
+
+        result.terms[action_ab] = (
+            result.terms.get(action_ab, 0.0) + coefficient)
+
+    # The other single pairing adds the mixed action b[0]^ a[1]
+    elif one_body_action_a[0][0] == one_body_action_b[1][0]:
+        action_ba = ((one_body_action_b[0][0], 1),
+                     (one_body_action_a[1][0], 0))
+
+        result.terms[action_ba] = (
+            result.terms.get(action_ba, 0.0) - coefficient)
+
+
+def _commutator_one_body_with_two_body(action_a, action_b,
+                                       coefficient, result):
+    """Compute commutator of action-specified one- and two-body operators.
+
+    Args:
+        action_a, action_b (tuple): single terms, one one-body and the other
+            two-body, from normal-ordered FermionOperators. It does not matter
+            which is one- or two-body so long as only one of each appears.
+        coefficient (complex float): coefficient of the commutator.
+        result (FermionOperator): the resultant to which to add the commutator.
```

Comment:
I still think it is strange there is an input argument named "result". Do you really get a performance advantage from performing the addition within the function?

Reply:
Yeah, see e.g. https://stackoverflow.com/questions/16256913/improving-performance-of-very-large-dictionary-in-python for a bit more info - basically as the underlying dict for result gets bigger and bigger, operations start to slow down a bit. There are some ways around this but they aren't really applicable for us (since we don't have a good idea about the keys ahead of time). And that post is discussing the slowdown for dictionaries with ~100K elements - the commutators here can end up with hundreds of millions of terms, where cutting the time on dictionary operations by a factor 2-3 can make a noticeable difference.

Reply:
But how does adding directly to dictionary for the whole system, as opposed to creating a new dictionary that exists just within the scope of this function, help things?

Reply:
OK, I see what you're saying. If we compute a Trotter error operator using this function, we're computing a sum of double commutators of terms as they appear in the simulation. Having `prior_terms` means we can add directly to a single FermionOperator instead of having to total a large number of (individually large) FermionOperators. This is where having `prior_terms` reduces the number of operations.

Reply:
I ran a small benchmark computing the Trotter error operator for 8x8 spinless jellium. Using the code with the prior result (calling the function with `prior_terms=error_operator`) is a third faster than the identical code using `prior_terms += function()`. So the difference isn't trivial (not just a fraction of a percent ;)) but isn't gigantic either. I'd vote to leave it in, but I'm fine removing it if you think it's much cleaner without.

Reply:
Ready for final review (changed to `prior_terms`)! My commit marked this thread outdated but it has the final point for discussion I think - thoughts @babbush @kevinsung ?

Reply:
I guess it is fine to leave in. This PR is good to go from my perspective. Just waiting on @kevinsung now.